packages/indexer/src/transaction/watcher.ts

```diff
@@ -91,154 +132,156 @@ class TransactionWatcher {
-        // We can't set the checked up to slot as the current tx's slot, since there might be a tx after this one on the same slot. So we instead
-        // need to set it to the prior tx's slot if that slot is less than the current slot.
-        const newCheckedUpToSlot =
-          slot > priorSlot ? priorSlot : this.checkedUpToSlot;
-        const updateResult = await db.con
-          .update(schema.transactionWatchers)
-          .set({
-            acct,
-            latestTxSig: signature,
-            firstTxSig: curWatcherRecord.firstTxSig ?? signature,
-            serializerLogicVersion: SERIALIZED_TRANSACTION_LOGIC_VERSION,
-            checkedUpToSlot: newCheckedUpToSlot,
-          })
-          .where(eq(schema.transactionWatchers.acct, acct))
-          .returning({ acct: schema.transactionWatchers.acct });
-        if (updateResult.length !== 1 || updateResult[0].acct !== acct) {
-          logger.info(
-            `Failed to update tx watcher for acct ${acct} on tx ${signature}`
+          .returning({ txSig: schema.transactions.txSig });
+        if (upsertResult.length !== 1 || upsertResult[0].txSig !== signature) {
+          logger.error(
+            `Failed to upsert ${signature}. ${JSON.stringify(
+              transactionRecord
+            )}`
           );
-          process.exit(1);
-        }
-        priorSlot = slot;
-        this.checkedUpToSlot = newCheckedUpToSlot;
-        numIndexed++;
-        if (numIndexed % 50 === 0) {
-          logger.info(`(${numIndexed} / ${history.length}) ${acct} watcher`);
+          return Err({ type: WatcherBackfillError.TransactionUpsertFailure });
         }
       }
-      // Update checkedUpToSlot to latest confirmed slot. If we don't do this, then checkedUpToSlot would only be updated once there's a new
-      // transaction, and this would mean indexers would stall in cases where some of the dependent watchers don't have frequent transactions.
-      // It's possible that this might be the source of bugs if somehow new transactions come in that are before the latest confirmed slot but were
-      // not returned by the RPC's tx history.
-      // EDIT: encountered above bug so trying to resolve by switching to finalized slot rather than confirmed.
-      const newCheckedUpToSlot = this.checkedUpToSlot > latestFinalizedSlot ? this.checkedUpToSlot : latestFinalizedSlot;
-      if (newCheckedUpToSlot === latestFinalizedSlot) {
-        console.log(`For acct ${acct}, using finalized slot of ${latestFinalizedSlot}`);
-      }
-      const updateResult = await db.con.update(schema.transactionWatchers)
+      // TODO: maybe i need to validate below succeeded. I can't use returning because this isn't an upsert so it could
+      //       be a no-op in the happy path
+      await db.con
+        .insert(schema.transactionWatcherTransactions)
+        .values({
+          txSig: signature,
+          slot,
+          watcherAcct: acct,
+        })
+        .onConflictDoNothing();
```

Comment:
Be careful with this and postgres, it maintains an AI so if there is an ID with AI it'll continue to increase regardless.

Reply:
There's no AI columns in the entire schema